substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/MemoryUtil.java

```diff
@@ -81,54 +87,77 @@ public static List<String> determineMemoryFlags() {
+        reasonableMaxMemorySize = Math.min(reasonableMaxMemorySize, maxAllowedMemorySize);
+
         return (double) reasonableMaxMemorySize / totalMemorySize * 100;
     }
 
     /**
-     * Returns the total amount of available memory in bytes on Linux based on
-     * <code>/proc/meminfo</code>, otherwise <code>-1</code>. Note that this metric is not
-     * container-aware (does not take cgroups into account) and may report available memory of the
-     * host.
+     * Returns the total amount of available memory plus headroom in bytes on Linux based on cgroup
+     * metrics or <code>/proc/meminfo</code>, otherwise <code>-1</code>.
      *
      * @see <a href=
      *      "https://github.com/torvalds/linux/blob/865fdb08197e657c59e74a35fa32362b12397f58/mm/page_alloc.c#L5137">page_alloc.c#L5137</a>
      */
     private static long getAvailableMemorySizeLinux() {
+        var containerMetrics = jdk.internal.platform.Container.metrics();
+        if (containerMetrics != null) {
+            long memoryLimit = containerMetrics.getMemoryLimit();
+            if (memoryLimit >= 0) {
+                /*
+                 * In containerized environments with a memory limit, do not use any headroom to
+                 * avoid OutOfMemoryErrors.
+                 */
+                return (long) (memoryLimit * MAX_ALLOWED_CONTAINER_MEMORY_RATIO);
+            }
+        }
+
         try {
-            String memAvailableLine = Files.readAllLines(Paths.get("/proc/meminfo")).stream().filter(l -> l.startsWith("MemAvailable")).findFirst().orElse("");
-            Matcher m = Pattern.compile("^MemAvailable:\\s+(\\d+) kB").matcher(memAvailableLine);
-            if (m.matches()) {
-                return Long.parseLong(m.group(1)) * KiB_TO_BYTES;
+            List<String> memInfo = Files.readAllLines(Paths.get("/proc/meminfo"));
+            String swapTotalLine = memInfo.stream().filter(l -> l.startsWith("SwapTotal")).findFirst().orElse("");
+            Matcher swapTotalMatcher = Pattern.compile("^SwapTotal:\\s+(\\d+) kB").matcher(swapTotalLine);
+            long swapTotalKiB = swapTotalMatcher.matches() ? Long.valueOf(swapTotalMatcher.group(1)) : 0;
+            String memAvailableLine = memInfo.stream().filter(l -> l.startsWith("MemAvailable")).findFirst().orElse("");
+            Matcher memAvailableMatcher = Pattern.compile("^MemAvailable:\\s+(\\d+) kB").matcher(memAvailableLine);
+            if (memAvailableMatcher.matches()) {
+                long memAvailableKiB = Long.parseLong(memAvailableMatcher.group(1));
+                /*
+                 * Based on total swap, set headroom somewhere between zero (no swap) and
+                 * MAX_HEADROOM_RATIO of available memory (total swap exceeds available memory).
+                 */
+                long reasonableHeadroomKiB = (long) (Math.min(swapTotalKiB, memAvailableKiB) * MAX_HEADROOM_RATIO);
+                return (memAvailableKiB + reasonableHeadroomKiB) * KiB_TO_BYTES;
```

Comment:
> I wonder what the headroom is supposed to achieve.

It's supposed to compensate for the cases when available memory at startup is lower than it could be. Also, there's no easy way to determine swap space on macOS and Windows, so this helps there too.

> Considering a patch which used 50% of total swap space as the extra headroom.

How about using 80% of free swap space, or 20% of available memory, whichever is larger? If we use 50% of total swap space but the swap if full, aren't we going to put the system in trouble?

Reply:
> > I wonder what the headroom is supposed to achieve.
> 
> It's supposed to compensate for the cases when available memory at startup is lower than it could be.

Got it. It makes me think if the current patch will help (on Linux), though. I welcome the container simplification.

> 
> > Considering a patch which used 50% of total swap space as the extra headroom.
> 
> How about using 80% of _free_ swap space, or 20% of available memory, whichever is larger?

I think that would be better than the current approach.

> If we use 50% of total swap space but the swap if full, aren't we going to put the system in trouble?

Note that the previous heuristic before considering available memory (only) - current master - didn't consider whether or not swap was available. It used 80% of physical memory, AFAIK. Considering a system with 64GB of memory and 32 gb of swap, it would use up to ~51gb of memory. On systems with little available memory at the time, it's conceivable that would cause swapping, possibly up to 100% of total swap. Now we are on the other extreme of the pendulum since only available memory (estimate before using swap) is being used. Given that, I find 50% of total swap isn't unreasonable as an upper bound to land somewhere between where we were and current master. That's my thinking.

Using the available swap metric has it's own set of problems as it's again a snapshot value at a specific point in time.

Anyway, I think the higher of '80% of free swap and 20% of available memory' will do the trick too.

Reply:
> Now we are on the other extreme of the pendulum since only available memory (estimate before using swap) is being used. Given that, I find 50% of total swap isn't unreasonable as an upper bound to land somewhere between where we were and current master. That's my thinking.

Yes, you are right and I agree. I'm just not sure if we want to go straight to the middle of the two approaches. My thinking was to try a 20% headroom first before going to 50% of swap.

> Anyway, I think the higher of '80% of free swap and 20% of available memory' will do the trick too.

Ok, let's go with this for now. Keep in mind that we still make sure that the MaxRAMPercentage is about 100%, and that we use more than 32GB of memory to avoid compressed pointers. So we should be fine if the headroom exceeds any of the two upper limits.